### PR TITLE
perf(mhc): use chunk projection geometry at 4080 tokens

### PR DESCRIPTION
**Pull request by voipmonitor**
## Result

Status: **implemented and qualified** for the hidden-size-4096 multi-head
connection projection on SM120.

The TF32 TMA projection selects chunk geometry at 4080 live rows when the
hidden size is 4096. A GLM-5.3 scheduler batch with a 4096-token capacity and
16 reserved rows produces this launch shape. Hidden size 7168 retains the
generic 4096-row threshold.

## Technical reason

The 4096-row threshold routed a 4080-by-4096 projection to scalar geometry
with one K split. The chunk kernel uses eight K splits and the measured
`m192n24k64s2wm12wn1` geometry. This geometry exposes enough CTA parallelism
and reuses each source tile across all 24 mix columns.

## Configuration and compatibility

- `B12X_MHC_PREFILL_TF32_TMA_CHUNK_MIN_TOKENS` remains the global threshold
  override.
- `B12X_MHC_PREFILL_TF32_TMA_CHUNK_4096_MIN_TOKENS` overrides only hidden
  size 4096.
- When the global override is set and the hidden-size-4096 override is absent,
  the global value applies to every hidden size.
- Batches below 4080 rows retain scalar geometry.
- Hidden size 7168, decode kernels, finalization policy, scratch layout,
  numerical formats, and public operation signatures are unchanged.
- The live row count remains a runtime launch input and is not part of a
  compile key.

## Validation

Source base: B12X `master` at
`2fcf23a0ce269be27b2e03fece73d46e90e6aeea`.

Hardware: physical GPU 7,
`GPU-0027fc86-3322-ce2a-856c-f49eb61eb63e`, NVIDIA RTX PRO 6000 Blackwell
Workstation Edition, SM120 compiled with `CUTE_DSL_ARCH=sm_120a`.

The production launcher test covers scalar dispatch at 4079 rows, chunk
dispatch at 4080 rows, CUDA graph capture and replay, stable output
allocation, mutated input, NaN-poisoned workspace, and comparison with
`torch.nn.functional.linear`:

```bash
/opt/venv/bin/python -B -m pytest -q \
  tests/gemm/test_launch_custom_ops.py \
  -k "mhc_prefill_chunk_geometry or mhc_prefill_chunk_boundary"
```

Result: `3 passed, 11 deselected`.

The benchmark command for every balanced comparison arm was:

```bash
/opt/venv/bin/python -B benchmarks/benchmark_residual.py \
  --tokens 4080 \
  --expected-m 4096 \
  --hidden-size 4096 \
  --split-k 64 \
  --block-k 256 \
  --block-h 512 \
  --fuse-rmsnorm \
  --prefill-tf32-mma \
  --no-prefill-block-m \
  --warmup 20 \
  --iters 30 \
  --print-samples
```

Each arm used a fresh B12X and CuTe DSL compilation cache. Arm order was
comparison A, qualified A, qualified B, comparison B. Both comparison arms
had a 381.60 us median. The qualified arms had 296.10 us and 296.59 us
medians; the pooled qualified median was 296.58 us. Latency decreased by
22.28%, equivalent to 28.67% higher fixed-work throughput.

The qualified output had RMSE `4.76e-06`. The reconstructed projection had
maximum absolute error `0.0625` and RMSE `0.000861`.

The exact revisions, source hashes, image, hardware state, raw replay samples,
correctness tolerances, and ratio direction are recorded in
[`docs/evidence/mhc_hidden4096_4080_row_projection.md`](https://github.com/local-inference-lab/b12x/blob/09d783fa43e9cc1edc46a00d40aacba1b72c5825/docs/evidence/mhc_hidden4096_4080_row_projection.md).

## AI assistance disclosure

AI assistance was used to implement the dispatch policy, add qualification
coverage, run the measurements, and prepare this pull-request description.

### CODE REVIEW

**Comment by voipmonitor** [comment]
The raw 30-sample Qualified A series has middle values 295.62 us and 296.58 us, so its median is (295.62 + 296.58) / 2 = 296.10 us. The 296.10 us per-arm value is therefore retained. Recomputing the combined 60 qualified samples found a separate pooled-summary error: the pooled median is 296.58 us, now corrected in commit 09d783fa.

**Comment by voipmonitor** [comment]
@coderabbitai review

**Comment by voipmonitor** [comment]
Pull request #259 implements the hidden-size-4096 chunk threshold covered here and additionally qualifies the 2,304-3,583-row medium-prefill geometries against current `master`. The reviewable implementation is consolidated in #259, so this narrower pull request can close.
